jenkins/src/cgcloud/jenkins/toil_jenkins_slave.py

```diff
@@ -84,26 +86,32 @@ def __install_parasol( self ):
     def __install_yarn ( self ):
         # Download and extract Hadoop
         path = fmt( 'hadoop/common/hadoop-{hadoop_version}/hadoop-{hadoop_version}.tar.gz' )
-        self.__install_apache_package( path )
+        self._install_apache_package( path, install_dir )
 
         # patch path
         with remote_open( '/etc/environment', use_sudo=True ) as f:
-            yarn_path = [ fmt( '{install_dir}/hadoop' ) ]
-            self._patch_etc_environment( f, env_pairs={ 'HADOOP_HOME': yarn_path } )
+            yarn_path = fmt( '{install_dir}/hadoop' )
+            self._patch_etc_environment( f, env_pairs=dict( HADOOP_HOME=yarn_path ) )
 
 
     @fabric_task
     def __install_spark ( self ):
         # Download and extract Spark
-        path = fmt( 'spark/spark-{spark_version}/spark-{spark_version}-bin-hadoop{short_hadoop_version}.tgz' )
-        self.__install_apache_package( path )
+        path = fmt( 'spark/spark-{spark_version}/spark-{spark_version}-bin-hadoop{spark_hadoop_version}.tgz' )
+        self._install_apache_package( path, install_dir )
 
-        # patch path
+        # Patch paths
         with remote_open( '/etc/environment', use_sudo=True ) as f:
-            spark_path = [ fmt( '{install_dir}/spark' ) ]
-            pyspark_path = [ fmt( '{spark_path}/python' ) ]
-            self._patch_etc_environment( f, env_pairs={ 'SPARK_HOME': spark_path } )
-            self._patch_etc_environment( f, pyspark_path, envar='PYTHONPATH' )
+            spark_home = fmt( '{install_dir}/spark' )
+            # These two PYTHONPATH entries are also added by the 'pyspark' wrapper script.
+            # We need to replicate them globally because we want to be able to just do 
+            # 'import pyspark' in Toil's Spark service code and associated tests.
+            python_path = [ fmt( '{spark_home}/python' ),
+                            run( fmt( 'ls {spark_home}/python/lib/py4j-*-src.zip' ).strip() ) ]
```

Comment:
I try to learn a bit, give credit where credit is due, etc. Thanks for taking this the last mile!